src/common/tests/abt_perf.c

```diff
@@ -25,81 +25,115 @@
 #include <abt.h>
 #include <daos/common.h>
 #include <getopt.h>
+#include <time.h>
 
 static unsigned long	abt_cntr;
-static unsigned long	abt_created;
+static int		abt_ults;
 static bool		abt_waiting;
 static bool		abt_exiting;
 
 static ABT_pool		abt_pool;
 static ABT_cond		abt_cond;
 static ABT_mutex	abt_lock;
 static ABT_xstream	abt_xstream;
+static ABT_thread_attr	abt_attr = ABT_THREAD_ATTR_NULL;
+static char		*abt_name;
 
+static int		opt_concur = 1;
 static int		opt_secs;
-static int		opt_threads;
+static int		opt_stack;
+static int		opt_cr_type;
+
+static inline uint64_t
+abt_current_ms(void)
+{
+	struct timespec now;
+
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	return now.tv_sec * 1000 + now.tv_nsec / 1000000;
+}
 
 static void
 abt_thread_1(void *arg)
 {
 	ABT_mutex_lock(abt_lock);
-	abt_created--;
-	if (abt_waiting)
+	if (!abt_exiting && abt_ults < opt_concur) {
+		/* less than the threshold, try to create more at here */
+		abt_ults++;
+		abt_cntr++;
+		ABT_mutex_unlock(abt_lock);
+
+		ABT_thread_create(abt_pool, abt_thread_1, NULL, abt_attr, NULL);
+
+		ABT_mutex_lock(abt_lock);
+	} /* else: do nothing and exit */
+
+	abt_ults--;
+	if (abt_waiting) {
 		ABT_cond_broadcast(abt_cond);
+		abt_waiting = false;
+	}
 	ABT_mutex_unlock(abt_lock);
 }
 
 /**
  * Create UTLs for a few seconds, concurrent creation is always lower
- * than @opt_threads.
+ * than @opt_concur.
  */
 static void
-abt_create_rate(void)
+abt_ult_create_rate(void)
 {
-	double		then;
-	double		now;
-	double		prt;
+	uint64_t	then;
+	uint64_t	now;
+	uint64_t	prt;
 	int		rc;
+	int		nsec = 0;
 
-	prt = then = ABT_get_wtime();
+	prt = now = then = abt_current_ms();
 	while (1) {
 		if (!abt_exiting) {
-			now = ABT_get_wtime();
-			if (now - then >= opt_secs)
+			now = abt_current_ms();
+			if (now - then >= opt_secs * 1000)
```

Comment:
minor, some time helpers in gurt can use, for example d_timeus_secdiff(opt_secs).